ouroboros-network-framework/src/Ouroboros/Network/Snocket.hs

```diff
@@ -378,31 +357,58 @@ namedPipeSnocket iocp name = Snocket {
 -- Windows/POSIX type aliases
 --
 
-
+localSnocket :: AssociateWithIOCP -> FilePath -> LocalSnocket
 -- | System dependent LocalSnocket type
 #if defined(mingw32_HOST_OS)
 type LocalSnocket = HANDLESnocket
 type LocalFD      = Win32.HANDLE
 
-localSnocket :: AssociateWithIOCP -> FilePath -> LocalSnocket
 localSnocket = namedPipeSnocket
 #else
-type LocalSnocket = SocketSnocket
+type LocalSnocket = Snocket IO Socket LocalAddress
 type LocalFD      = Socket
 
-localSnocket :: AssociateWithIOCP -> FilePath -> LocalSnocket
-localSnocket iocp _  = rawSocketSnocket iocp
-#endif
+localSnocket iocp _ = Snocket {
+      getLocalAddr  = fmap toLocalAddress . Socket.getSocketName
+    , getRemoteAddr = fmap toLocalAddress . Socket.getPeerName
+    , addrFamily    = const LocalFamily
+    , connect       = \s addr -> do
+        Socket.connect s (fromLocalAddress addr)
+    , bind          = \fd addr -> Socket.bind fd (fromLocalAddress addr)
+    , listen        = flip Socket.listen 1
+    , accept        = fmapAccept toLocalAddress . (berkeleyAccept iocp)
+    , open          = openSocket
+    , openToConnect = \addr -> openSocket LocalFamily
+    , close         = Socket.close
+    , toBearer      = Mx.socketAsMuxBearer
+    }
+  where
+    toLocalAddress :: SockAddr -> LocalAddress
+    toLocalAddress (SockAddrUnix path) = LocalAddress path
+    toLocalAddress _                   = error "localSnocket.toLocalAddr: impossible happend"
 
-#if defined(mingw32_HOST_OS)
-type LocalAddress = FilePath
-#else
-type LocalAddress = Socket.SockAddr
+    fromLocalAddress :: LocalAddress -> SockAddr
+    fromLocalAddress = SockAddrUnix . getFilePath
+
+    openSocket :: AddressFamily LocalAddress -> IO Socket
+    openSocket LocalFamily = do
+      sd <- Socket.socket AF_UNIX Socket.Stream Socket.defaultProtocol
+      associateWithIOCP iocp (Right sd)
+        -- open is designed to be used in `bracket`, and thus it's called with
+        -- async exceptions masked.  The 'associteWithIOCP' is a blocking
+        -- operation and thus it may throw.
+        `catch` \(e :: IOException) -> do
+          Socket.close sd
+          throwIO e
+        `catch` \(SomeAsyncException _) -> do
+          Socket.close sd
+          throwIO e
+      return sd
 #endif
 
 localAddressFromPath :: FilePath -> LocalAddress
 #if defined(mingw32_HOST_OS)
-localAddressFromPath = id
+localAddressFromPath = LocalAddress
 #else
-localAddressFromPath = Socket.SockAddrUnix
+localAddressFromPath = LocalAddress
```

Comment:
Unix and windows case look identical with this change, do you still need `#if`?

Reply:
Indeed, looks like the `#if` can go.

Reply:
I'll remove it